README.md
# Zy

> Minimal and blazing-fast file server. For real, this time.

[![Crates.io](https://img.shields.io/crates/v/zy?label=latest)](https://crates.io/crates/zy)
[![CI](https://github.com/miraclx/zy/actions/workflows/ci.yml/badge.svg?branch=master)](https://github.com/miraclx/zy/actions/workflows/ci.yml)
[![MIT or Apache 2.0 Licensed](https://img.shields.io/crates/l/zy.svg)](#license)

## Features

- [Single Page Application support](https://developer.mozilla.org/en-US/docs/Glossary/SPA)
- Partial responses ([Range](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range) support)
- Cross-Origin Resource Sharing ([CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS))
- [Automatic HTTP compression](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding) (Zstd, Gzip, Brotli, Deflate)
- Dynamic [cache control](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control) (ETag, Last-Modified, Cache-Control)
- Auto-served `index.html` files
- Sane defaults
  - No access to hidden files
  - No access to content outside the base directory
  - No access to symbolic links outside the base directory

## Installation

You can download any of the pre-compiled binaries from the [releases page](https://github.com/miraclx/zy/releases).

Or if you already have Rust installed, you can install it with `cargo`:

> - Please, note that the minimum supported version of Rust for `zy` is `1.59.0`.
> - Also, that the binary may be bigger than expected because it contains debug symbols. This is intentional. To remove debug symbols and therefore reduce the file size, you can instead run it with the `--profile slim` or simply just run `strip` on it.

```console
cargo install zy
```

Alternatively, you can also build the latest version of `zy` directly from GitHub.

```console
cargo install --git https://github.com/miraclx/zy.git
```

## Usage

```console
zy
```

_This will start serving your current directory on <http://localhost:3000> by default._

_...you can also specify a different port or base directory:_

```console
zy /path/to/serve
```

_...or perhaps different addresses:_

```console
zy -l 5000 -l 127.0.0.1:8080 -l 192.168.1.25
```

## Configuration

You can run `zy --help` to see all available options.

```console
$ zy --help
Zy 0.2.0
Minimal and blazing-fast file server.

USAGE:
    zy [OPTIONS] [DIR]

ARGS:
    <DIR>    Directory to serve [default: .]

OPTIONS:
    -l, --listen <URI>    Sets the address to listen on (repeatable) [default: 127.0.0.1:3000]
                          Valid: `3000`, `127.0.0.1`, `127.0.0.1:3000` [env: PORT]
    -s, --spa             Run as a Single Page Application
    -i, --index <FILE>    Index file to serve from the base directory [default: index.html]
        --404 <FILE>      404 file to serve from the base directory [default: 404.html]
    -c, --cache <TIME>    Cache time (max-age) [default: 1h]
                          Valid: `10` for 10 seconds, `1h`, `1year 6months`
        --no-cors         Disable Cross-Origin Resource Sharing (CORS)
    -a, --all             Serve hidden files
    -f, --follow-links    Follow symlinks outside of the base directory (unsafe)
    -v, --verbose         Be verbose
    -x, --confirm-exit    Require confirmation before exiting on Ctrl+C
    -Z, --anonymize       Hide the `Server` and `X-Powered-By` headers [alias: `--anon`]
    -h, --help            Print help information
    -V, --version         Print version information
```

## Credits

Zy was originally inspired by [sfz](https://github.com/weihanglo/sfz), [serve](https://github.com/vercel/serve) and [http-server](https://github.com/http-party/http-server). It is written in [Rust](https://rust-lang.org) and uses [actix](https://github.com/actix/actix-web) as the web framework.

## Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall be
dual licensed as below, without any additional terms or conditions.

## License

Licensed under either of

- Apache License, Version 2.0
   ([LICENSE-APACHE](LICENSE-APACHE) or <http://www.apache.org/licenses/LICENSE-2.0>)
- MIT license
   ([LICENSE-MIT](LICENSE-MIT) or <http://opensource.org/licenses/MIT>)

at your option.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.0] - 2022-11-06

- Support for the `PORT` environment variable.
- Human cache time input (e.g. `1h`, `1year 6months`).
- `--anonymize` flag to hide the `Server` and `X-Powered-By` headers.
- Added `zstd` compression support.
- Dynamic cache control (ETag, Last-Modified, Cache-Control).
- Auto-served `index.html` files.
- Use `println` over tracing for trivial logs.

## [0.1.1] - 2022-10-30

> Release Page: <https://github.com/miraclx/zy/releases/tag/v0.1.1>

[unreleased]: https://github.com/miraclx/zy/compare/v0.2.0...HEAD
[0.2.0]: https://github.com/miraclx/zy/compare/v0.1.1...v0.2.0
[0.1.1]: https://github.com/miraclx/zy/releases/tag/v0.1.1


src/exit.rs
use std::future::Future;

use color_eyre::Result;
use tokio::signal;
use tracing::info;

/// A future that never resolves, but will call the handler when it detects a shutdown signal.
///
/// Signals: `SIGINT`, `SIGTERM`, `SIGHUP` or explicit `shutdown` call.
///
/// A use-case of this, is to explicitly shutdown the server. Which then cancels this future.
pub async fn on_signal<F, Fut>(confirm_exit: bool, handler: F) -> Result<()>
where
    F: FnOnce(bool) -> Fut,
    Fut: Future<Output = ()>,
{
    let sigint = async {
        if confirm_exit {
            let mut last_signal_timestamp = None;

            loop {
                signal::ctrl_c().await?;

                let now = std::time::Instant::now();
                if let Some(last_signal_timestamp) = last_signal_timestamp {
                    if now.duration_since(last_signal_timestamp) < std::time::Duration::from_secs(5)
                    {
                        info!("[signal] Ctrl-C received");
                        break;
                    }
                }
                info!("[signal] Ctrl-C received, press again to exit");
                last_signal_timestamp = Some(now);
            }
        } else {
            signal::ctrl_c().await?;
        }

        Result::<()>::Ok(())
    };

    let sigterm = async {
        #[cfg(unix)]
        {
            let mut sigterm = signal::unix::signal(signal::unix::SignalKind::terminate())?;
            sigterm.recv().await;
            info!("[signal] SIGTERM received");
        }

        #[cfg(windows)]
        {
            let mut sigterm = signal::windows::ctrl_break()?;
            sigterm.recv().await;
            info!("[signal] Ctrl-Break received");
        }

        Result::<()>::Ok(())
    };

    let sighup = async {
        #[cfg(unix)]
        {
            let mut sighup = signal::unix::signal(signal::unix::SignalKind::hangup())?;
            sighup.recv().await;
            info!("[signal] SIGHUP received");
        }

        #[cfg(windows)]
        std::future::pending::<()>().await;

        Result::<()>::Ok(())
    };

    let res = tokio::select! {
        _ = sighup => handler(false),
        _ = sigint => handler(false),
        _ = sigterm => handler(true),
    };

    tokio::select! {
        _ = async {
            res.await;
            std::future::pending::<()>().await;
        } => {},
        _ = signal::ctrl_c() => {
            info!("[signal] Ctrl-C received while exiting, forcibly exiting...");
        }
    }

    Ok(())
}


src/main.rs
use std::env;
use std::ffi::OsStr;
use std::io;
use std::path::{Component, Path, PathBuf};
use std::process;
use std::sync::Arc;
use std::time;

use actix_web::http::{header, StatusCode};
use actix_web::{guard, web, App, HttpServer};
use actix_web::{HttpRequest, HttpResponse};
use clap::Parser;
use color_eyre::eyre::Result;
use humantime::format_duration;
use tracing::{debug, info};
use tracing_subscriber::EnvFilter;

mod cli;
mod exit;
mod middleware;
#[macro_use]
mod macros;

// /a => a
// a/../b => b
// /a/b/../c/./d => a/c/d
// C:\a => ERROR
fn normalize_path<P: AsRef<Path>>(path: P) -> io::Result<PathBuf> {
    let mut buf = PathBuf::new();
    for c in path.as_ref().components() {
        match c {
            Component::Normal(c) => buf.push(c),
            Component::ParentDir => {
                buf.pop();
            }
            Component::CurDir | Component::RootDir => {}
            _ => {
                return Err(io::Error::new(
                    io::ErrorKind::InvalidData,
                    "prefix is not supported",
                ))
            }
        }
    }
    for c in buf.components() {
        assert!(matches!(c, Component::Normal(_)));
    }
    Ok(buf)
}

#[derive(Copy, Clone)]
enum CachePolicy {
    NoCache,
    ShouldCache,
    Indeterminate,
}

#[derive(Copy, Clone)]
enum PathSource {
    Client,
    Server,
}

fn serve(
    req: &HttpRequest,
    path: &str,
    source: PathSource,
    state: &ServerState,
) -> Option<HttpResponse> {
    let path = normalize_path(Path::new(path)).ok()?;

    if let PathSource::Client = source {
        if !state.args.all
            && path
                .file_name()
                .map_or(false, |name| name.to_string_lossy().starts_with('.'))
        {
            return None;
        }
    }

    let mut path = state
        .args
        .dir
        .canonical
        .join(if path.as_os_str().is_empty() {
            state.args.dir.canonical.join(&state.args.index)
        } else {
            state.args.dir.canonical.join(path).canonicalize().ok()?
        });

    if let PathSource::Client = source {
        if !path.starts_with(&state.args.dir.canonical) && !state.args.follow_links {
            return None;
        }
    }

    if path.is_dir() {
        path = path.join("index.html").canonicalize().ok()?;
    }

    if !path.is_file() {
        return None;
    }

    if state.args.verbose {
        debug!(target: "zy::serve", path=%path.strip_prefix(&state.args.dir.canonical).ok()?.display());
    }

    let file = actix_files::NamedFile::open(&path).ok()?;

    let mut res = file
        .use_etag(true)
        .prefer_utf8(true)
        .use_last_modified(true)
        .disable_content_disposition()
        .into_response(req);

    // https://github.com/GoogleChrome/lighthouse/blob/60c2fa25d11187802e905e4f335b2e7f6df735f1/core/audits/byte-efficiency/uses-long-cache-ttl.js#L144-L164
    if let StatusCode::OK | StatusCode::PARTIAL_CONTENT = res.status() {
        let ext = path.extension().and_then(OsStr::to_str);
        let mime = ext.map_or(mime::APPLICATION_OCTET_STREAM, |ext| {
            actix_files::file_extension_to_mime(ext)
        });

        // https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/devtools/front_end/common/ResourceType.js;l=279-291;drc=3dab2a4e337404c9f853eccb07a0f647004319e2
        let cache_policy = match (mime.type_(), mime.subtype()) {
            (mime::TEXT, mime::HTML) => CachePolicy::NoCache,
            (mime::APPLICATION, mime::JAVASCRIPT)
            | (mime::TEXT, mime::CSS)
            | (mime::IMAGE, _)
            | (mime::TEXT, _)
            | (mime::FONT, _) => CachePolicy::ShouldCache,
            _ if matches!(ext, Some("otf") | Some("woff")) => CachePolicy::ShouldCache,
            _ => CachePolicy::Indeterminate,
        };

        let mut cache_directives = vec![];

        match cache_policy {
            CachePolicy::NoCache => {
                cache_directives.extend([
                    header::CacheDirective::NoCache,
                    header::CacheDirective::NoStore,
                ]);
            }
            CachePolicy::ShouldCache => {
                cache_directives.extend([
                    header::CacheDirective::Public,
                    header::CacheDirective::MaxAge(state.args.cache),
                ]);
            }
            CachePolicy::Indeterminate => {}
        }

        if let Ok((k, v)) =
            header::TryIntoHeaderPair::try_into_pair(header::CacheControl(cache_directives))
        {
            res.headers_mut().insert(k, v);
        }
    }

    Some(res)
}

async fn index(
    req: HttpRequest,
    path: web::Path<String>,
    state: web::Data<Arc<ServerState>>,
) -> HttpResponse {
    if state.args.verbose {
        debug!(
            target: "zy::request",
            version = ?req.version(),
            method = %req.method(),
            uri = %req.uri(),
        );
    }

    let mut res = match serve(&req, &path, PathSource::Client, &state) {
        Some(res) => res,
        None => {
            if state.args.spa {
                let accepts_html = <header::Accept as header::Header>::parse(&req)
                    .map_or(false, |accept| {
                        accept.iter().any(|mime| mime.item == "text/html")
                    });
                if accepts_html {
                    if state.args.verbose {
                        info!(target: "zy::serve", "SPA routing to {}", state.args.index);
                    }
                    if let Some(res) = serve(&req, &state.args.index, PathSource::Server, &state) {
                        return res;
                    }
                }
            }
            if state.args.verbose {
                info!(target: "zy::serve", "not found, serving {}", state.args.not_found);
            }
            match serve(&req, &state.args.not_found, PathSource::Server, &state) {
                Some(mut resp) => {
                    *resp.status_mut() = StatusCode::NOT_FOUND;
                    resp
                }
                None => {
                    if state.args.verbose {
                        info!(target: "zy::serve", "{} not found, omitting response body", state.args.not_found);
                    }
                    HttpResponse::build(StatusCode::NOT_FOUND).finish()
                }
            }
        }
    };

    if !state.args.no_cors {
        res.headers_mut().insert(
            header::ACCESS_CONTROL_ALLOW_ORIGIN,
            header::HeaderValue::from_static("*"),
        );
    }

    res
}

pub struct ServerState {
    args: cli::Args,
}

async fn init_app() -> Result<()> {
    let mut args = cli::Args::parse();

    if let Ok(port) = env::var("PORT") {
        if port.parse::<u16>().is_err() {
            eprintln!(
                "warning: invalid PORT environment variable: {:?}, ignoring",
                port
            );
        }
    }

    if args.listen.is_empty() {
        args.listen.push(cli::addr_from_str("127.0.0.1").unwrap());
    }

    print_block! {
        "Process ID" => process::id()
        "Base Directory" => if args.dir.is_current_dir() {
            format!("{} (current directory)", args.dir.raw.display())
        } else {
            args.dir.raw.display().to_string()
        }
        "Cache Time" => format_duration(time::Duration::from_secs(args.cache as u64))
        "Single Page App?" => args.spa
        "Index File" => args.index
        "404 Fallback" => args.not_found
        "Cross-Origin Resource Sharing" => if args.no_cors {
            "disabled"
        } else {
            "(any domain)"
        }
        "Anonymize Server" => args.anonymize
        "Will Serve Hidden Files" => args.all
        "Follow Symlinks Outside Root" => args.follow_links
        "Confirm Exit" => args.confirm_exit
        "Be Verbose" => args.verbose
    }

    let server_state = Arc::new(ServerState { args });

    let mut server = {
        let server_state = server_state.clone();
        HttpServer::new(move || {
            App::new()
                .wrap(middleware::ZyServer {
                    anonymize: server_state.args.anonymize,
                })
                .app_data(web::Data::new(server_state.clone()))
                .service(
                    web::resource("/{path:.*}")
                        .guard(guard::Any(guard::Get()).or(guard::Head()))
                        .wrap(middleware::Compress::default())
                        .to(index),
                )
        })
        .disable_signals()
    };

    for addr in &server_state.args.listen {
        server = server.bind(addr)?;
        println!("Listening on http://{}", addr);
    }

    let server = server.run();

    let server_handle = server.handle();

    tokio::select! {
        _ = server => {}
        _ = exit::on_signal(
            server_state.args.confirm_exit,
            |graceful| async move {
                if graceful {
                    info!("Starting graceful shutdown");
                } else {
                    info!("Shutting down immediately");
                }
                server_handle.stop(graceful).await;
            }
        ) => {}
    }

    Ok(())
}

fn setup() -> Result<()> {
    if std::env::var("RUST_LIB_BACKTRACE").is_err() {
        std::env::set_var("RUST_LIB_BACKTRACE", "1");
    }
    color_eyre::install()?;

    tracing_subscriber::fmt()
        .with_env_filter(
            EnvFilter::from_default_env()
                .add_directive("zy=info".parse()?)
                .add_directive("zy=debug".parse()?),
        )
        .init();

    Ok(())
}

#[actix_web::main]
async fn main() -> Result<()> {
    setup()?;

    init_app().await?;

    Ok(())
}


src/routes.rs
use std::sync::Arc;

use axum::body::Body;
use axum::routing::Router;

use super::ServerState;

mod service {
    use std::io;
    use std::sync::Arc;

    use axum::body::Body;
    use axum::extract::Extension;
    use axum::http::StatusCode;
    use axum::routing::{get, get_service, Router};
    use tower_http::compression::CompressionLayer;
    use tower_http::services::ServeDir;

    use super::ServerState;

    pub fn load(state: Arc<ServerState>) -> Router<Body> {
        let serve_dir_service =
            get_service(ServeDir::new(&state.dir)).handle_error(|error: io::Error| async move {
                (
                    StatusCode::INTERNAL_SERVER_ERROR,
                    format!("Unhandled internal error: {}", error),
                )
            });

        Router::new()
            .layer(CompressionLayer::new().gzip(true))
            .route("/ping", get(|| async { "pong" }))
            .fallback(serve_dir_service)
            .layer(Extension(state))
    }
}

mod handler {
    use std::sync::Arc;

    use axum::body::Body;
    use axum::extract::Extension;
    use axum::http::{Request, StatusCode};
    use axum::response::Response;
    use axum::routing::{get, Router};
    use tower::ServiceExt;
    use tower_http::compression::CompressionLayer;
    use tower_http::services::fs::ServeFileSystemResponseBody;
    use tower_http::services::ServeDir;

    use super::ServerState;

    pub fn load(state: Arc<ServerState>) -> Router<Body> {
        Router::new()
            .layer(CompressionLayer::new().gzip(true))
            .route("/ping", get(|| async { "pong" }))
            .fallback(Router::new().route("/*path", get(handler)))
            .layer(Extension(state))
    }

    async fn handler(
        Extension(state): Extension<Arc<ServerState>>,
        request: Request<Body>,
    ) -> Result<Response<ServeFileSystemResponseBody>, (StatusCode, String)> {
        ServeDir::new(&state.dir)
            .oneshot(request)
            .await
            .map_err(|err| {
                (
                    StatusCode::INTERNAL_SERVER_ERROR,
                    format!("Unhandled internal error: {}", err),
                )
            })
    }
}

pub fn load(state: Arc<ServerState>) -> Router<Body> {
    if true {
        service::load(state.clone())
    } else {
        handler::load(state.clone())
    }
}


src/middleware.rs
use std::future::{ready, Future, Ready};
use std::pin::Pin;
use std::task::{Context, Poll};

use actix_web::dev::{Service, ServiceRequest, ServiceResponse, Transform};
use actix_web::http::header;
pub use actix_web::middleware::Compress;

pub struct ZyServer {
    pub anonymize: bool,
}

impl<S, B> Transform<S, ServiceRequest> for ZyServer
where
    S: Service<ServiceRequest, Response = ServiceResponse<B>>,
    S::Future: 'static,
    B: 'static,
{
    type Response = S::Response;
    type Error = S::Error;
    type Transform = ZyServerMiddleware<S>;
    type InitError = ();
    type Future = Ready<Result<Self::Transform, Self::InitError>>;

    fn new_transform(&self, service: S) -> Self::Future {
        ready(Ok(ZyServerMiddleware {
            service,
            anonymize: self.anonymize,
        }))
    }
}

pub struct ZyServerMiddleware<S> {
    service: S,
    anonymize: bool,
}

impl<S, B> Service<ServiceRequest> for ZyServerMiddleware<S>
where
    S: Service<ServiceRequest, Response = ServiceResponse<B>>,
    S::Future: 'static,
    B: 'static,
{
    type Response = S::Response;
    type Error = S::Error;
    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>>>>;

    fn poll_ready(&self, ctx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        self.service.poll_ready(ctx)
    }

    fn call(&self, req: ServiceRequest) -> Self::Future {
        let fut = self.service.call(req);

        let anonymous = self.anonymize;
        Box::pin(async move {
            let mut res = fut.await?;

            if !anonymous {
                res.headers_mut().insert(
                    header::SERVER,
                    header::HeaderValue::from_static(concat!("Zy/", env!("CARGO_PKG_VERSION"))),
                );

                res.headers_mut().insert(
                    "X-Powered-By"
                        .try_into()
                        .expect("x-powered-by should be valid header"),
                    header::HeaderValue::from_static(concat!("Zy/", env!("CARGO_PKG_VERSION"))),
                );
            }

            Ok(res)
        })
    }
}


src/cli.rs
use std::env;
use std::ffi::OsStr;
use std::io;
use std::net::{AddrParseError, IpAddr, SocketAddr};
use std::num::{FpCategory, IntErrorKind};
use std::path::PathBuf;

use clap::{AppSettings, Parser};
use humantime::{parse_duration, DurationError};

const DEFAULT_PORT: u16 = 3000;

pub fn addr_from_str(s: &str) -> Result<SocketAddr, AddrParseError> {
    if let Ok(port) = s.parse::<u16>() {
        return Ok(SocketAddr::from(([127, 0, 0, 1], port)));
    }
    if let Ok(host) = s.parse::<IpAddr>() {
        return Ok(SocketAddr::from((
            host,
            env::var("PORT")
                .ok()
                .and_then(|p| p.parse().ok())
                .unwrap_or(DEFAULT_PORT),
        )));
    }
    s.parse::<SocketAddr>()
}

#[derive(Debug)]
pub struct CanonicalizedPath {
    pub raw: PathBuf,
    pub canonical: PathBuf,
}

impl CanonicalizedPath {
    pub fn is_current_dir(&self) -> bool {
        env::current_dir().map_or(false, |cwd| cwd == self.canonical)
    }
}

fn parse_canonicalize_dir(s: &OsStr) -> Result<CanonicalizedPath, io::Error> {
    let raw = PathBuf::from(s);
    let canonical = raw.canonicalize()?;
    Ok(CanonicalizedPath { raw, canonical })
}

fn parse_cache_time(s: &str) -> color_eyre::Result<u32> {
    let duration = match parse_duration(s) {
        Ok(duration) => duration,
        Err(err) => {
            if matches!(
                err,
                DurationError::UnknownUnit { ref unit, .. }
                if unit.is_empty()
            ) {
                match s.parse::<u32>() {
                    Ok(seconds) => return Ok(seconds),
                    Err(err) if *err.kind() == IntErrorKind::PosOverflow => {
                        return Err(color_eyre::eyre::eyre!("cache time is too large"))
                    }
                    Err(_) => {}
                }
            }
            return Err(err.into());
        }
    };
    let secs = duration.as_secs_f64();
    assert!(
        matches!(secs.classify(), FpCategory::Normal | FpCategory::Zero),
        "humantime should not return NaN or infinite values"
    );
    if secs != secs.trunc() {
        return Err(color_eyre::eyre::eyre!(
            "cache time in seconds cannot be fractional"
        ));
    }
    if secs > u32::MAX as f64 {
        return Err(color_eyre::eyre::eyre!("cache time is too large"));
    }
    Ok(secs as u32)
}

#[derive(Debug, Parser)]
#[clap(name = "Zy")]
#[clap(about, version, setting = AppSettings::DeriveDisplayOrder)]
pub struct Args {
    /// Directory to serve
    #[clap(default_value = ".", parse(try_from_os_str = parse_canonicalize_dir))]
    pub dir: CanonicalizedPath,

    /// Sets the address to listen on (repeatable) [default: 127.0.0.1:3000]
    /// Valid: `3000`, `127.0.0.1`, `127.0.0.1:3000` [env: PORT]
    #[clap(short, long, value_name = "URI", multiple_occurrences = true)]
    #[clap(verbatim_doc_comment, parse(try_from_str = addr_from_str))]
    pub listen: Vec<SocketAddr>,

    /// Run as a Single Page Application
    #[clap(short, long)]
    pub spa: bool,

    /// Index file to serve from the base directory.
    #[clap(short, long, value_name = "FILE", default_value = "index.html")]
    pub index: String,

    /// 404 file to serve from the base directory.
    #[clap(long = "404", value_name = "FILE", default_value = "404.html")]
    pub not_found: String,

    /// Cache time (max-age) [default: 1h]
    /// Valid: `10` for 10 seconds, `1h`, `1year 6months`
    #[clap(short, long, value_name = "TIME", verbatim_doc_comment)]
    #[clap(default_value = "1h", hide_default_value = true)]
    #[clap(parse(try_from_str = parse_cache_time))]
    pub cache: u32,

    /// Disable Cross-Origin Resource Sharing (CORS)
    #[clap(long)]
    pub no_cors: bool,

    /// Serve hidden files
    #[clap(short, long)]
    pub all: bool,

    /// Follow symlinks outside of the base directory (unsafe)
    #[clap(short, long)]
    pub follow_links: bool,

    /// Be verbose
    #[clap(short, long)]
    pub verbose: bool,

    /// Require confirmation before exiting on Ctrl+C
    #[clap(short = 'x', long)]
    pub confirm_exit: bool,

    /// Hide the `Server` and `X-Powered-By` headers [alias: `--anon`]
    #[clap(short = 'Z', long, alias = "anon")]
    pub anonymize: bool,
}
